File: min_In_Rotated_SortedArray.java
//Time Complexity : O(logn)
//Space Complexity : O(1)
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : Found difficulty in finding a the best solution at first attempt.

class min_In_Rotated_SortedArray {
	
	public int findMin(int[] nums) {

		int leftIndex = 0;
		int rightIndex = nums.length-1;
		
		if(nums == null || nums.length == 0)
			return Integer.MAX_VALUE;

		while(rightIndex - leftIndex > 1)
		{
			int midValue = leftIndex + (rightIndex - leftIndex) / 2;
			
			if(nums[rightIndex] < nums[midValue])
				leftIndex = midValue + 1;
			else if(nums[leftIndex] < nums[midValue])
				rightIndex = midValue - 1;
			else
				rightIndex = midValue;
        }
        if(nums[leftIndex] < nums[rightIndex])
			return nums[leftIndex];
		else
			return nums[rightIndex];
			
		
	}
	
	public static void main(String[] args) { 
		int[] nums = {4,5,6,7,-1,0,1,2,3};
		int[] nums2 = {3,4,5,1,2};
		
		min_In_Rotated_SortedArray srch = new min_In_Rotated_SortedArray();
		
		System.out.println(srch.findMin(nums));
		
		System.out.println(srch.findMin(nums2));

		
	}
}

File: minInRotatedSortedArray.go
// Time Complexity : O(log(n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/*
 * Approach:
 * Since it is a rotated sorted array, at least one half will be sorted.
 * Another observation is that the minimum element is either at the unsorted
 * side while searching, or else at the start of any subarray.
 */

package main

func findMin(nums []int) int {
	if len(nums) == 0 {
		return -1
	}

	l, h := 0, len(nums)-1

	for l <= h {
		if nums[l] <= nums[h] { // if sorted, directly return first elem
			return nums[l]
		}

		m := l + (h-l)/2 // prevent integer overflow

		if (m == l || nums[m] < nums[m-1]) &&
			(m == h || nums[m] < nums[m+1]) {
			return nums[m] // you're at min since neighbors are bigger
		}

		if nums[l] <= nums[m] { // left is sorted, move right
			l = m + 1
		} else { // right is sorted, move left
			h = m - 1
		}
	}

	return -1
}
